TwoPageEBookReader/app/src/main/java/com/example/ebook_reader_sample/BookActivity.kt

```diff
@@ -0,0 +1,258 @@
+
+        setSupportActionBar(findViewById(R.id.toolbar))
+        supportActionBar?.title = intent.getStringExtra("BOOK_TITLE")
+        supportActionBar?.setDisplayHomeAsUpEnabled(true)
+
+        bookPagerView = findViewById(R.id.view_pager_view)
+        bookPagerView.viewTreeObserver.addOnGlobalLayoutListener(this)
+    }
+
+    //onGlobalLayout() gets the layout metrics of the empty pager view, and passes them to the Book object
+    //At the end, renderBook() is called
+    override fun onGlobalLayout() {
+        bookPagerView.viewTreeObserver.removeOnGlobalLayoutListener(this)
+
+        val bookPagerLocation = IntArray(2)
+        bookPagerView.getLocationInWindow(bookPagerLocation)
+        val bookPagerRect = Rect(bookPagerLocation[0], bookPagerLocation[1], bookPagerLocation[0] + bookPagerView.width, bookPagerLocation[1] + bookPagerView.height)
+        val captionHeight = resources.getDimension(R.dimen.caption_height).toInt()
+
+        val tempPageRects = ArrayList<Rect>()
+        when (layoutStateContainer.layoutMode) {
+            LayoutMode.NORMAL -> {
+                tempPageRects.add(Rect(bookPagerRect.left, bookPagerRect.top, bookPagerRect.right, bookPagerRect.bottom - captionHeight))
+            }
+            //TODO Setting book to two-page layout
+            LayoutMode.SPLIT_VERTICAL -> {
+                tempPageRects.add(Rect(bookPagerRect.left, bookPagerRect.top, bookPagerRect.right, layoutStateContainer.foldRect.top))
+                tempPageRects.add(Rect(bookPagerRect.left, layoutStateContainer.foldRect.bottom, bookPagerRect.right, bookPagerRect.bottom - captionHeight))
+            }
+            LayoutMode.SPLIT_HORIZONTAL -> {
+                tempPageRects.add(Rect(bookPagerRect.left, bookPagerRect.top, layoutStateContainer.foldRect.left, bookPagerRect.bottom - captionHeight))
+                tempPageRects.add(Rect(layoutStateContainer.foldRect.right, bookPagerRect.top, bookPagerRect.right, bookPagerRect.bottom - captionHeight))
+            }
+        }
+
+        book.pageRects = tempPageRects
+
+        renderBook()
+    }
+
+    //renderBook() creates or replaces the pager in the pager view
+    private fun renderBook() {
+        if (bookPagerView.adapter != null) {
+            bookPagerView.unregisterOnPageChangeCallback(pagePagerCallback)
+        }
+
+        bookPagerView.adapter = BookPagerAdapter(book, layoutStateContainer)
+        bookPagerView.setPageTransformer(PageTransformer())
+        val position = if (layoutStateContainer.layoutMode == LayoutMode.NORMAL) book.currentPage + 1 else (book.currentPage/2) + 1
+        bookPagerView.setCurrentItem(position, false)
+        bookPagerView.registerOnPageChangeCallback(pagePagerCallback)
+        bookPagerView.isUserInputEnabled = true
+    }
+
+    inner class BookPagerCallback() : ViewPager2.OnPageChangeCallback() {
+        override fun onPageSelected(position: Int) {
+            when (position) {
+                0 -> {
```

Comment:
Is nicer that you move `when` individual cases to small functions so you can read it like:
```
when (position) {
  0 -> doThis()
  1 -> doThat()
  else -> otherwise()
}
```

Reply:
Moved to smaller functions